jlm/hls/backend/rvsdg2rhls/rhls-dne.cpp

```diff
@@ -267,6 +269,159 @@ dead_loop(rvsdg::Node * ndmux_node)
   return true;
 }
 
+bool
+dead_loop_lcb(rvsdg::Node * lcb_node)
+{
+  auto lcb_op = dynamic_cast<const hls::loop_constant_buffer_op *>(&lcb_node->GetOperation());
+  JLM_ASSERT(lcb_op);
+  // one branch
+  if (lcb_node->output(0)->nusers() != 1)
+  {
+    return false;
+  }
+  auto branch_in = dynamic_cast<jlm::rvsdg::node_input *>(*lcb_node->output(0)->begin());
+  auto bo = dynamic_cast<const branch_op *>(&branch_in->node()->GetOperation());
+  if (!branch_in || !bo || !bo->loop)
+  {
+    return false;
+  }
+  // no user
+  if (branch_in->node()->output(1)->nusers())
+  {
+    return false;
+  }
+  // depend on same control
+  auto branch_cond_origin = branch_in->node()->input(0)->origin();
+  auto pred_buf_out = dynamic_cast<jlm::rvsdg::node_output *>(lcb_node->input(0)->origin());
+  if (!pred_buf_out
+      || !dynamic_cast<const predicate_buffer_op *>(&pred_buf_out->node()->GetOperation()))
+  {
+    return false;
+  }
+  auto pred_buf_cond_origin = pred_buf_out->node()->input(0)->origin();
+  // TODO: remove this once predicate buffers decouple combinatorial loops
+  auto extra_buf_out = dynamic_cast<jlm::rvsdg::node_output *>(pred_buf_cond_origin);
+  if (!extra_buf_out || !dynamic_cast<const buffer_op *>(&extra_buf_out->node()->GetOperation()))
+  {
+    return false;
+  }
+  auto extra_buf_cond_origin = extra_buf_out->node()->input(0)->origin();
+
+  if (auto pred_be = dynamic_cast<backedge_argument *>(extra_buf_cond_origin))
+  {
+    extra_buf_cond_origin = pred_be->result()->origin();
+  }
+  if (extra_buf_cond_origin != branch_cond_origin)
+  {
+    return false;
+  }
+  // divert users
+  branch_in->node()->output(0)->divert_users(lcb_node->input(1)->origin());
+  remove(branch_in->node());
+  remove(lcb_node);
+  return true;
+}
+
+bool
+fix_mem_split(rvsdg::Node * split_node)
```

Comment:
Combine multiple memory state split nodes into a single node to reduce HW.
Removes unused split outputs, which are a result from running fix_mem_merge first.

Reply:
Should likely be part of a node reduction pass insteast of DNE